metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -2103,6 +2122,21 @@ class MetalsLanguageServer(
     }
   }
 
+  @JsonRequest("workspace/willRenameFiles")
+  def willRenameFiles(
+      params: RenameFilesParams
+  ): CompletableFuture[WorkspaceEdit] =
+    CancelTokens.future { _ =>
+      val moves = params.getFiles.asScala.toSeq.map { rename =>
+        packageProvider.willMovePath(
+          rename.getOldUri().toAbsolutePath,
+          rename.getNewUri().toAbsolutePath,
+          referencesProvider,
+        )
+      }
+      Future.sequence(moves).map(_.mergeChanges)
```

Comment:
Could we deduplicate the changes when analyzing the file? Is this actually still needed?

Reply:
If the implementation of `willMovePath` is correct, there should be no duplicates at this point. 
`mergeChanges` serves two purposes here:
1. Merging: `moves` is a collection of WorkspaceEdit's, so they need to be merged anyway into a single WorkspaceEdit. `willMovePath` might return a `Map` with changes, but merging would be still needed
2. Resiliency: if there is some corner-case resulting in duplicates, reporting them as-is to VSCode will result in an error (generally this error occurs in case there are overlapping ranges in a single workspace edit

Reply:
Ok cool, thanks for the explanation!